### PR TITLE
Conformance code improvements

### PR DESCRIPTION
**Pull request by shrutiparabgoogle**
Changes in this PR:
* Added a check to match `mime_type` while deciding which styleguides to apply.
* Refactor conformance code:
   - Replaced the use of maps to a struct which holds all the metadata required to run the linter and compute the report.
   - Modularized the code and split the functions into `linter` and `conformance-task` specific go files. This will help us in writing better unit tests for specific functions.
   - Created another folder `cmd/registry/conformance` which holds the above two files in package `conformance`.
   - This new package is imported by `compute/conformance.go` to make use of the core conformance logic.

Next items to follow:
* More tests for `linter.go` and `conformance-task.go`.

### CODE REVIEW

File: cmd/registry/conformance/conformance-task.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+				// Check if the guideline report for the guideline which contains this rule
+				// has already been initialized. If it hasn't then create one.
+				guidelineReport, ok := guidelineReportsMap[guideline.GetId()]
+				if !ok {
+					guidelineReport = initializeGuidelineReport(guideline.GetId())
+
+					guidelineReportsMap[guideline.GetId()] = guidelineReport
+					conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports =
+						append(conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports, guidelineReport)
```
Comment (seaneganx) [inline cmd/registry/conformance/conformance-task.go:175]: Can this be split into smaller steps? There's a lot going on in this one statement. Maybe assigning the group to a variable before appending to its GuidelineReports slice would simplify it a good amount.

File: cmd/registry/conformance/linter.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	"time"
+
+	"github.com/apigee/registry/log"
+	"github.com/apigee/registry/rpc"
+	"google.golang.org/protobuf/proto"
+)
+
+type ruleMetadata struct {
+	guidelineRule *rpc.Rule      //Rule object associated with the linter-rule
+	guideline     *rpc.Guideline //Guideline object associated with the linter-rule
```
Comment (seaneganx) [inline cmd/registry/conformance/linter.go:32]: These comments don't look like the others in this file. Specifically 1) missing the space after `//` and 2) not written as a full sentence with a period at the end.

**Comment by shrutiparabgoogle** [comment]
> LGTM. There's a lot to go through here so I only skimmed the PR and commented about a few nitpicks that popped out at me. Let me know if you want me to do a deeper review.

Thanks @seaneganx for the review! Most of the change in this PR includes moving code from the main `conformance.go` file into separate files. I will be adding unit tests for these new files, where the original code might go through some refactor, that might be a good time to do a deep review.  I wanted you to review this change so that you are aware of the refactoring.

I will implement your suggestions in this PR.

File: cmd/registry/conformance/linter.go (hunk 2 above)
Comment (shrutiparabgoogle) [reply cmd/registry/conformance/linter.go]: Done

File: cmd/registry/conformance/conformance-task.go (hunk 1 above)
Comment (shrutiparabgoogle) [reply cmd/registry/conformance/conformance-task.go]: Done

File: cmd/registry/conformance/conformance-task.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+func (task *ComputeConformanceTask) String() string {
+	return fmt.Sprintf("compute %s", conformanceReportName(task.Spec.GetName(), task.StyleguideId))
+}
+
+func (task *ComputeConformanceTask) Run(ctx context.Context) error {
+	log.Debugf(ctx, "Computing conformance report %s", conformanceReportName(task.Spec.GetName(), task.StyleguideId))
+
+	// Download spec
+
+	data, err := core.GetBytesForSpec(ctx, task.Client, task.Spec)
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/conformance/conformance-task.go]: Done

File: cmd/registry/conformance/conformance-task.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	"google.golang.org/protobuf/proto"
+)
+
+func conformanceReportName(specName string, styleguideName string) string {
+	return fmt.Sprintf("%s/artifacts/conformance-%s", specName, styleguideName)
+}
+
+func initializeConformanceReport(
+	specName string,
+	styleguideId string) *rpc.ConformanceReport {
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/conformance/conformance-task.go]: Done

File: cmd/registry/cmd/compute/conformance.go [hunk 5]
```diff
@@ -98,294 +81,42 @@ func conformanceCommand(ctx context.Context) *cobra.Command {
...
-					guidelineIdToGuidelineReport[guideline.GetId()] = report
-
-					conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports =
-						append(conformanceReport.GuidelineReportGroups[guideline.Status].GuidelineReports, report)
+	err = core.ListSpecs(ctx, client, spec, filter, func(spec *rpc.ApiSpec) {
+		// Check if the styleguide definition contains the mime_type of the spec
+		for _, supportedType := range styleguide.GetMimeTypes() {
+			if supportedType == spec.GetMimeType() {
+				// Delegate the task of computing the conformance report for this spec
+				// to the worker pool.
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/compute/conformance.go]: Done

File: cmd/registry/cmd/compute/conformance.go [hunk 6]
```diff
@@ -54,42 +44,35 @@ func conformanceCommand(ctx context.Context) *cobra.Command {
...
-				if err != nil {
-					log.FromContext(ctx).WithError(err).Debugf("Failed to get message type for MIME type %q", artifact.GetMimeType())
-					return
-				}
-
-				if messageType != "google.cloud.apigeeregistry.applications.v1alpha1.StyleGuide" {
-					// Ignore any artifact that isn't a style guide.
-					return
-				}
+			artifactFilter := "mime_type.contains('google.cloud.apigeeregistry.applications.v1alpha1.StyleGuide')"
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/compute/conformance.go]: Done

**Review by timburks — APPROVED** [review]

**Pull request merged**